stratum/hal/lib/barefoot/bfrt_table_manager.cc

```diff
@@ -678,6 +704,195 @@ ::util::Status BfrtTableManager::ReadTableEntry(
+  ASSIGN_OR_RETURN(
+      ::p4::v1::TableEntry result,
+      BuildP4TableEntry(table_entry, table_key.get(), table_data.get()));
+  // Build sugared version of action profile.
+  ::p4::v1::ActionProfileActionSet* action_profile_set =
+      result.mutable_action()->mutable_action_profile_action_set();
+  uint64 action_profile_group_id;
+  RETURN_IF_ERROR(table_data->GetSelectorGroupId(&action_profile_group_id));
+  std::vector<int> member_ids;
+  std::vector<std::unique_ptr<BfSdeInterface::TableDataInterface>> action_datas;
+  RETURN_IF_ERROR(bf_sde_interface_->GetActionProfileMembers(
+      device_, session, bfrt_act_prof_table_id, action_profile_group_id,
+      &member_ids, &action_datas));
+  for (const auto& action_data : action_datas) {
+    auto action = action_profile_set->add_action_profile_actions();
+    ASSIGN_OR_RETURN(
+        ::p4::v1::TableEntry member_action,
+        BuildP4TableEntry(table_entry, table_key.get(), action_data.get()));
+    *action->mutable_action() = member_action.action().action();
+    action->set_weight(1);
+  }
+  ::p4::v1::ReadResponse resp;
+  *resp.add_entities()->mutable_table_entry() = result;
+  VLOG(1) << "ReadOneShotActionProfile resp " << resp.DebugString();
+  if (!writer->Write(resp)) {
+    return MAKE_ERROR(ERR_INTERNAL) << "Write to stream for failed.";
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status BfrtTableManager::WriteOneShotActionProfile(
+    std::shared_ptr<BfSdeInterface::SessionInterface> session,
+    const ::p4::v1::Update::Type type,
+    const ::p4::v1::TableEntry& table_entry) {
+  ASSIGN_OR_RETURN(auto table,
+                   p4_info_manager_->FindTableByID(table_entry.table_id()));
+  ASSIGN_OR_RETURN(auto action_profile, p4_info_manager_->FindActionProfileByID(
+                                            table.implementation_id()));
+  ASSIGN_OR_RETURN(uint32 table_id,
+                   bf_sde_interface_->GetBfRtId(table_entry.table_id()));
+  ASSIGN_OR_RETURN(
+      uint32 bfrt_act_prof_table_id,
+      bf_sde_interface_->GetBfRtId(action_profile.preamble().id()));
+  ASSIGN_OR_RETURN(
+      uint32 bfrt_act_sel_table_id,
+      bf_sde_interface_->GetActionSelectorBfRtId(bfrt_act_prof_table_id));
+
+  // Ensure that one-shot and traditional programming style are not mixed on the
+  // same action selector or table.
+  bool uses_one_shot_action =
+      gtl::LookupOrInsert(&p4_entity_uses_one_shot_action_map_,
+                          action_profile.preamble().id(), true) &&
+      gtl::LookupOrInsert(&p4_entity_uses_one_shot_action_map_,
+                          table_entry.table_id(), true);
+  if (!uses_one_shot_action) {
+    RETURN_ERROR(ERR_INVALID_PARAM)
+        << "One-shot and traditional action profile programming styles "
+           "cannot be mixed on the same action profile "
+        << action_profile.preamble().name() << " or table "
+        << table_entry.table_id() << ".";
+  }
+
+  if (type == ::p4::v1::Update::DELETE) {
```

Comment:
I think modify is a superset of delete and insert, the logic should be here already.